### DISCUSSION TITLE
pyright and pyright-langserver gives different results?

### DISCUSSION THREAD

**Discussion by ebetica** [microsoft/pyright#1635] [Q&A] [answered]
Hi! I'm trying to interface pyright with Kakoune, since it seems to be significantly faster than pyls. However, I'm having a problem with import resolution.

Here's a minimal working example:

```
$ tree
.
├── main.py
├── setup.py
└── testpkg
    ├── __init__.py
    └── pkg.py

1 directory, 4 files
$ cat main.py 
from testpkg import pkg # There is a problem in this line.

pkg.f()
$ cat testpkg/pkg.py 
def f():
    print("HI!")
```

First, I run `pyright main.py`, which claims there are no issues, and the import goes through. Then, I run it through the language server via Kakoune's kak-lsp package. I see that a `textDocument/didOpen` was sent to the server, with the correct uri (`file:///path/to/main.py`). The server claims no source files are found, which I think is correct since it's using `pyright-langserver --stdio` to communicate. However, it's incorrectly saying that it cannot import this package.

I'm not sure if this helps at all, but the language server also never displays that it found `pyrightconfig.json` if I include one, whereas `pyright` will find the configuration. Is there some other path I should be passing into the language server to get imports correct? Does it work in other editors?

**Comment by erictraut** [comment]
The pyright language server has been integrated successfully with neovim, sublime and emacs. You can find links to these projects on [the main pyright page](https://github.com/microsoft/pyright#installation). The authors of these projects may be able to help you debug the issue you're hitting.

My guess is that the workspace root directory is not being passed through correctly from the editor to the LSP.

**Reply by ebetica** [reply]
I see that workspace/didChangeConfiguration is sent an empty json. Are the fields suppose to line up with the json config described here? https://github.com/microsoft/pyright/blob/master/docs/configuration.md Thanks for the pointers, I can try asking some of these authors as well.

**Reply by jakebailey** [reply]
An "empty" didChangeConfiguration is actually a notification from the client to the server to say "you need to query me for the info", as pyright uses the pull model for settings rather than asking for full config events. The message that contains the workspace root is the "initialize" request, which is always the first one.

**Reply by erictraut** [reply]
It's fine for didChangeConfiguration to have an empty json.

I suspect that the problem is with the initialize message. It should be passed a set of parameters that includes either a `workspaceFolders` or `rootPath` field. The former is used for multi-root workspaces (if the editor supports that). The latter is used for simple single-root workspaces.

**Reply by ebetica** [reply]
Ah - thanks a lot! We are sending a `rootUri` field, which looks to be deprecated in favor of `workspaceFolders`. I'm guessing pyright somehow doesn't read this deprecated field then?

**Reply by erictraut** [reply]
Both `rootPath` and `rootUri` are deprecated. Pyright will use `rootPath` if it's provided, but it doesn't do anything with `rootUri`. So that explains the problem you're seeing.

**Reply by jakebailey** [reply]
MPLS used rootUri if present, then rootPath (predating workspaceFolders; URI replaced the path IIRC); perhaps we can just take it and use it if it's a valid `file://` URI?

**Reply by erictraut** [reply]
I don't think there's a need to change it. Both fields are deprecated at this point. We will continue to support `rootPath` for backward compatibility.

**Comment by ebetica** [comment, marked as answer]
TL;DR: pyright supports `rootPath` but not `rootUri`, both are deprecated so `workspaceFolders` is what should be used going forwards.
